File: AIMDS/src/detection/normalize.ts
/**
 * Text normalisation for injection detection.
 *
 * Produces a canonical form (NFKC, zero-width stripped, confusables folded,
 * whitespace collapsed) plus a small set of bounded variants that undo common
 * obfuscations (separator-joined words, single-character spacing, leetspeak,
 * whitespace removal). Every transform is linear in input length.
 */

/** Zero-width and invisible formatting code points that carry no text. */
const INVISIBLE = /[\u200B-\u200F\u202A-\u202E\u2060-\u2064\uFEFF\u00AD\u180E]/g;

/**
 * Explicit confusables map: Cyrillic and Greek letters whose glyphs are
 * indistinguishable from Latin ones. NFKC does not fold these.
 */
const CONFUSABLES: Record<string, string> = {
  'а': 'a', 'А': 'A', // а А
  'е': 'e', 'Е': 'E', // е Е
  'о': 'o', 'О': 'O', // о О
  'р': 'p', 'Р': 'P', // р Р
  'с': 'c', 'С': 'C', // с С
  'у': 'y', 'У': 'Y', // у У
  'х': 'x', 'Х': 'X', // х Х
  'і': 'i', 'І': 'I', // і І
  'ј': 'j', 'Ј': 'J', // ј Ј
  'һ': 'h', 'Һ': 'H', // һ Һ
  'ԁ': 'd', 'ԛ': 'q', 'ԝ': 'w',
  'ѕ': 's', 'Ѕ': 'S', // ѕ Ѕ
  'в': 'B', 'В': 'B', // в В (visually B in caps)
  'к': 'k', 'К': 'K', // к К
  'м': 'm', 'М': 'M', // м М
  'н': 'H', 'Н': 'H', // н Н
  'т': 'T', 'Т': 'T', // т Т
  'α': 'a', 'Α': 'A', // α Α
  'ε': 'e', 'Ε': 'E', // ε Ε
  'ο': 'o', 'Ο': 'O', // ο Ο
  'ρ': 'p', 'Ρ': 'P', // ρ Ρ
  'ι': 'i', 'Ι': 'I', // ι Ι
  'κ': 'k', 'Κ': 'K', // κ Κ
  'ν': 'v', 'Ν': 'N', // ν Ν
  'τ': 't', 'Τ': 'T', // τ Τ
  'υ': 'u', 'Υ': 'Y', // υ Υ
  'χ': 'x', 'Χ': 'X', // χ Χ
  'Β': 'B', 'Η': 'H', 'Μ': 'M', 'Ζ': 'Z',
};

const CONFUSABLE_RE = new RegExp(`[${Object.keys(CONFUSABLES).join('')}]`, 'g');

/** Leetspeak digits/symbols that stand in for letters. */
const LEET: Record<string, string> = {
  '0': 'o', '1': 'i', '3': 'e', '4': 'a', '5': 's', '7': 't', '@': 'a', '$': 's', '!': 'i', '|': 'l',
};

export interface TextVariant {
  /** Name of the transform that produced this text. */
  readonly name: 'base' | 'separators' | 'leet' | 'compact';
  readonly text: string;
  /** True when the variant has no whitespace left, so `\s+`-free regexes must be used. */
  readonly compactForm: boolean;
}

/**
 * Canonical form used for all matching. Idempotent.
 */
export function normalizeText(input: string): string {
  return input
    .normalize('NFKC')
    .replace(INVISIBLE, '')
    .replace(CONFUSABLE_RE, (ch) => CONFUSABLES[ch] ?? ch)
    .replace(/[ \t\f\v ]+/g, ' ')
    .replace(/ ?\r?\n ?/g, '\n')
    .trim();
}

/**
 * Join runs of single characters separated by one repeated separator
 * ("i.g.n.o.r.e", "i g n o r e") into a word, and turn `_`/`-` between
 * letters into spaces ("ignore_all_previous" -> "ignore all previous").
 */
function undoSeparators(text: string): string {
  // Runs of 3+ single letters joined by the same separator: i.g.n / i g n / i-g-n
  const joined = text.replace(
    /\b(?:[A-Za-z][.\-_ ]){2,}[A-Za-z]\b/g,
    (run) => run.replace(/[.\-_ ]/g, ''),
  );
  return joined.replace(/(?<=[A-Za-z])[_\-]+(?=[A-Za-z])/g, ' ');
}

/**
 * Fold leet digits to letters, only inside tokens that already contain a
 * letter (so numbers, hashes, and IDs are left alone).
 */
function undoLeet(text: string): string {
  return text.replace(/[A-Za-z0-9@$!|]*[A-Za-z][A-Za-z0-9@$!|]*/g, (token) => {
    if (!/[0-9@$!|]/.test(token)) return token;
    return token.replace(/[0-9@$!|]/g, (ch) => LEET[ch] ?? ch);
  });
}

/**
 * Variants of a normalised text, de-duplicated. `base` is always first.
 * Variants are only generated for the first `limit` characters so the
 * scan cost stays bounded.
 */
export function textVariants(normalized: string, limit = 65536): TextVariant[] {
  const head = normalized.length > limit ? normalized.slice(0, limit) : normalized;
  const out: TextVariant[] = [{ name: 'base', text: normalized, compactForm: false }];
  const seen = new Set<string>([normalized]);
  const push = (name: TextVariant['name'], text: string) => {
    if (!seen.has(text)) {
      seen.add(text);
      out.push({ name, text, compactForm: name === 'compact' || !/\s/.test(text) });
    }
  };
  const sep = undoSeparators(head);
  push('separators', sep);
  push('leet', undoLeet(sep));
  push('compact', sep.replace(/[ \t.\-_]+/g, ''));
  return out;
}

/** Compact-form regex source: `\s+`/`\s*` become empty so joined words match. */
export function compactSource(source: string): string {
  return source.replace(/\\s[+*]/g, '');
}

File: AIMDS/crates/aidefence-core/src/normalize.rs
//! Text normalisation and obfuscation variants, mirroring `AIMDS/src/detection/normalize.ts`.
//!
//! `normalize_text` is the canonical form used for all matching: NFKC, invisible
//! characters stripped, Cyrillic/Greek confusables folded, horizontal whitespace
//! collapsed, newlines kept. `text_variants` adds bounded variants that undo common
//! obfuscations (separator-joined words, leetspeak, whitespace removal).
//!
//! `sanitize` is the Rust-only forwarding helper: control characters removed,
//! invisibles stripped, confusables folded, PII masked, iterated to a fixpoint.

use regex::Regex;
use std::sync::OnceLock;
use unicode_normalization::UnicodeNormalization;

/// Zero-width and invisible formatting code points that carry no text
/// (same class as the TypeScript `INVISIBLE` regex).
pub fn is_invisible(c: char) -> bool {
    matches!(
        c,
        '\u{200B}'..='\u{200F}'
            | '\u{202A}'..='\u{202E}'
            | '\u{2060}'..='\u{2064}'
            | '\u{FEFF}'
            | '\u{00AD}'
            | '\u{180E}'
    )
}

/// Explicit confusables map, identical to the TypeScript `CONFUSABLES` table
/// (including its upper-casing of lower-case `в`, `н`, `т`).
pub fn fold_confusable(c: char) -> char {
    match c {
        'а' => 'a',
        'А' => 'A',
        'е' => 'e',
        'Е' => 'E',
        'о' => 'o',
        'О' => 'O',
        'р' => 'p',
        'Р' => 'P',
        'с' => 'c',
        'С' => 'C',
        'у' => 'y',
        'У' => 'Y',
        'х' => 'x',
        'Х' => 'X',
        'і' => 'i',
        'І' => 'I',
        'ј' => 'j',
        'Ј' => 'J',
        'һ' => 'h',
        'Һ' => 'H',
        'ԁ' => 'd',
        'ԛ' => 'q',
        'ԝ' => 'w',
        'ѕ' => 's',
        'Ѕ' => 'S',
        'в' => 'B',
        'В' => 'B',
        'к' => 'k',
        'К' => 'K',
        'м' => 'm',
        'М' => 'M',
        'н' => 'H',
        'Н' => 'H',
        'т' => 'T',
        'Т' => 'T',
        'α' => 'a',
        'Α' => 'A',
        'ε' => 'e',
        'Ε' => 'E',
        'ο' => 'o',
        'Ο' => 'O',
        'ρ' => 'p',
        'Ρ' => 'P',
        'ι' => 'i',
        'Ι' => 'I',
        'κ' => 'k',
        'Κ' => 'K',
        'ν' => 'v',
        'Ν' => 'N',
        'τ' => 't',
        'Τ' => 'T',
        'υ' => 'u',
        'Υ' => 'Y',
        'χ' => 'x',
        'Χ' => 'X',
        'Β' => 'B',
        'Η' => 'H',
        'Μ' => 'M',
        'Ζ' => 'Z',
        other => other,
    }
}

fn leet(c: char) -> Option<char> {
    match c {
        '0' => Some('o'),
        '1' => Some('i'),
        '3' => Some('e'),
        '4' => Some('a'),
        '5' => Some('s'),
        '7' => Some('t'),
        '@' => Some('a'),
        '$' => Some('s'),
        '!' => Some('i'),
        '|' => Some('l'),
        _ => None,
    }
}

fn horizontal_ws_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new("[ \t\u{000C}\u{000B}\u{00A0}]+").expect("static regex"))
}

fn newline_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(" ?\r?\n ?").expect("static regex"))
}

fn joined_letters_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r"(?-u:\b)(?:[A-Za-z][.\-_ ]){2,}[A-Za-z](?-u:\b)").expect("static regex")
    })
}

fn leet_token_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[A-Za-z0-9@$!|]+").expect("static regex"))
}

fn compact_sep_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[ \t.\-_]+").expect("static regex"))
}

/// Canonical form used for all matching (mirrors `normalizeText`). Idempotent.
pub fn normalize_text(input: &str) -> String {
    let folded: String = input
        .nfkc()
        .filter(|c| !is_invisible(*c))
        .map(fold_confusable)
        .collect();
    let spaced = horizontal_ws_re().replace_all(&folded, " ");
    let lined = newline_re().replace_all(&spaced, "\n");
    lined.trim().to_string()
}

/// Join runs of single letters separated by one repeated separator
/// (`i.g.n.o.r.e`, `i g n o r e`) into a word, then turn `_` / `-` runs between
/// letters into spaces (`ignore_all_previous` becomes `ignore all previous`).
pub fn undo_separators(text: &str) -> String {
    let joined = joined_letters_re().replace_all(text, |caps: &regex::Captures<'_>| {
        caps[0]
            .chars()
            .filter(|c| !matches!(c, '.' | '-' | '_' | ' '))
            .collect::<String>()
    });
    let chars: Vec<char> = joined.chars().collect();
    let mut out = String::with_capacity(joined.len());
    let mut i = 0;
    while i < chars.len() {
        let c = chars[i];
        if matches!(c, '_' | '-') && i > 0 && chars[i - 1].is_ascii_alphabetic() {
            let mut j = i;
            while j < chars.len() && matches!(chars[j], '_' | '-') {
                j += 1;
            }
            if j < chars.len() && chars[j].is_ascii_alphabetic() {
                out.push(' ');
                i = j;
                continue;
            }
        }
        out.push(c);
        i += 1;
    }
    out
}

/// Fold leet digits / symbols to letters inside tokens that already contain a letter.
pub fn undo_leet(text: &str) -> String {
    leet_token_re()
        .replace_all(text, |caps: &regex::Captures<'_>| {
            let token = &caps[0];
            let has_letter = token.chars().any(|c| c.is_ascii_alphabetic());
            let has_leet = token.chars().any(|c| leet(c).is_some());
            if has_letter && has_leet {
                token.chars().map(|c| leet(c).unwrap_or(c)).collect()
            } else {
                token.to_string()
            }
        })
        .into_owned()
}

/// A text variant produced by [`text_variants`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TextVariant {
    /// `base`, `separators`, `leet` or `compact`.
    pub name: &'static str,
    /// The text.
    pub text: String,
    /// `true` when the text has no whitespace, so the compact regexes must be used.
    pub compact_form: bool,
}

/// Variants of a normalized text, de-duplicated; `base` is always first. Variants
/// are generated from the first `limit` characters so the scan cost stays bounded.
pub fn text_variants(normalized: &str, limit: usize) -> Vec<TextVariant> {
    let head: String = normalized.chars().take(limit).collect();
    let mut out = vec![TextVariant {
        name: "base",
        text: normalized.to_string(),
        compact_form: false,
    }];
    let push = |name: &'static str, text: String, out: &mut Vec<TextVariant>| {
        if out.iter().any(|v| v.text == text) {
            return;
        }
        let compact_form = name == "compact" || !text.chars().any(char::is_whitespace);
        out.push(TextVariant {
            name,
            text,
            compact_form,
        });
    };
    let sep = undo_separators(&head);
    let leet = undo_leet(&sep);
    let compact = compact_sep_re().replace_all(&sep, "").into_owned();
    push("separators", sep, &mut out);
    push("leet", leet, &mut out);
    push("compact", compact, &mut out);
    out
}

/// One sanitize pass: NFKC, invisibles and non-layout control characters removed,
/// confusables folded, NFKC again so folded letters recompose with marks, PII masked.
fn sanitize_once(text: &str) -> String {
    let stripped: String = text.chars().filter(|c| !is_invisible(*c)).collect();
    let folded: String = stripped
        .nfkc()
        .filter(|c| !is_invisible(*c))
        .filter(|c| !c.is_control() || matches!(c, '\n' | '\t' | '\r'))
        .map(fold_confusable)
        .collect();
    let recomposed: String = folded.nfkc().collect();
    crate::pii::mask_all(&recomposed)
}

/// Sanitize text for forwarding. Iterated to a fixpoint so
/// `sanitize(sanitize(x)) == sanitize(x)`.
pub fn sanitize(text: &str) -> String {
    let mut current = sanitize_once(text);
    for _ in 0..4 {
        let next = sanitize_once(&current);
        if next == current {
            break;
        }
        current = next;
    }
    current
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn folds_zero_width_fullwidth_and_cyrillic() {
        assert_eq!(
            normalize_text("ig\u{200B}nore ｉｇｎｏｒｅ іgnore"),
            "ignore ignore ignore"
        );
        assert_eq!(normalize_text("  a \t b \n c  "), "a b\nc");
    }

    #[test]
    fn undoes_separators_and_leet() {
        assert_eq!(
            undo_separators("i.g.n.o.r.e a_l_l ignore_all-previous"),
            "ignoreall ignore all previous"
        );
        assert_eq!(
            undo_leet("1gn0re all pr3vious 1nstructions 2024 a|b"),
            "ignore all previous instructions 2024 alb"
        );
    }

    #[test]
    fn produces_separator_leet_and_compact_variants() {
        let names: Vec<&str> = text_variants(
            &normalize_text("1gn0re_all previous i.n.s.t.r.u.c.t.i.o.n.s"),
            65536,
        )
        .iter()
        .map(|v| v.name)
        .collect();
        assert_eq!(names[0], "base");
        assert!(names.contains(&"separators"));
        assert!(names.contains(&"leet"));
        assert!(names.contains(&"compact"));
    }

    #[test]
    fn sanitize_keeps_newlines_and_drops_other_controls() {
        assert_eq!(sanitize("a\u{0007}b\nc\td"), "ab\nc\td");
    }

    #[test]
    fn sanitize_is_idempotent_on_combining_sequences() {
        let s = "а\u{0301} e\u{200D}\u{0301}";
        let once = sanitize(s);
        assert_eq!(sanitize(&once), once);
    }
}
